### PR TITLE
GitHub workflows / Dynamic Rollout

### PR DESCRIPTION
**Pull request by DanilBaibak**
# GitHub workflows / Dynamic Rollout

[That is the issue](https://github.com/pytorch/test-infra/issues/5132) that can be used to control access to ARC runners. You only need to adjust the first comment accordingly. Currently, we support such cases:
* Add a GitHub username to use ARC runners.
* Add "*" at the beginning to switch ALL users to ARC runners.
* Add "!" at the beginning to switch ALL users to old runners.
* By default we will use the old runners for the `main`, `nightly`, `release` and `landchecks` branches.

There is also an option to run both runners at the same time:
* To use both runners, add a second comment with the word "both".
* If we want to use only one type of runners, just remove the second comment.

### CODE REVIEW

**Comment by seemethere** [comment]
How does this change now that @ZainRizvi's PR is in to run both ARC and the old infra in parallel?:
* https://github.com/pytorch/pytorch/pull/124753

File: .github/workflows/check-user.yml
Comment (ZainRizvi) [inline .github/workflows/check-user.yml:38]: See how you can modify this curl command to use `secrets.GITHUB_TOKEN` as an authentication mechanism.  

Github heavily rate limits requests that lack authentication (on a per-ip address basis) and without that auth there's a good chance we'll be affected

File: .github/workflows/check-user.yml
Comment (ZainRizvi) [inline .github/workflows/check-user.yml:28]: Might be better to put this run logic inside a python script. 
Way more versatile and easier to manage that way

**Review by ZainRizvi — CHANGES_REQUESTED** [review]

**Review by ZainRizvi — CHANGES_REQUESTED** [review]

File: .github/workflows/check-user.yml
Comment (DanilBaibak) [reply .github/workflows/check-user.yml:38]: Ohhh, this is a very good input!

File: .github/workflows/check-user.yml
Comment (DanilBaibak) [reply .github/workflows/check-user.yml:28]: I thought about this. So far logic is pretty simple. If the solution becomes more cumbersome, I'll migrate it to a Python script.

File: .github/workflows/pull.yml
Comment (DanilBaibak) [reply .github/workflows/pull.yml]: I updated it to "${{ github.head_ref || github.ref_name }}". But it is hard to say what value will be for a cron job.

File: .github/workflows/_check-user.yml
Comment (jeanschmidt) [inline .github/workflows/_check-user.yml:21]: lets be very mindful of job names and file names. `check-user` is not clear what this job is actually doing.

Maybe something like `runner determinator` would make more sense

File: .github/workflows/_check-user.yml
Comment (jeanschmidt) [inline .github/workflows/_check-user.yml:18]: I believe that it should output both the `workflow-type` and `runner`. So we can use it to easily migrate jobs from one place to another. Dynamically.

Makes more sense to do it in this way and maintain the flexibility to quickly migrate workflows in the future.

File: .github/workflows/_check-user.yml
Comment (ZainRizvi) [inline .github/workflows/_check-user.yml:42]: Is the `tr -d '\r\n'` necessary?

File: .github/workflows/_check-user.yml
Comment (ZainRizvi) [inline .github/workflows/_check-user.yml:41]: Let's do two things for safety
1. Make this action have a safe fall back (to classic-infra) only if either the curl request or parsing fails
2. Update the issue to point to this action so that future maintainers can easily find where it's used

File: .github/workflows/_check-user.yml
Comment (ZainRizvi) [inline .github/workflows/_check-user.yml:49]: How do you feel about turning the comment into properly formatted yaml or json?

That way you can cleanly put everything in one block without having to worry as much about parsing special symbols

It'll also let us gradually shift from "always running in both locations" to "start opting people into ARC-only mode"

Eg. the comment block could look like:

```
Modes:
  - type: both
     users: # list of users opted into both
        - atalman
        - clee2000
        - ...
  - type: arc-only
     users: # users opted into  ARC-only
     - ZainRizvi
     - DanilBaibak
  - type: classic-only
     users:
     - * # default result

```

File: .github/workflows/_check-user.yml
Comment (ZainRizvi) [inline .github/workflows/_check-user.yml:60]: can this be made case insensitive? That's bitten us in the past occasionally

**Comment by ZainRizvi** [comment]
> How does this change now that @ZainRizvi's PR is in to run both ARC and the old infra in parallel?:
> 
> * [Get ARC jobs to run on both classic and ARC infra #124753](https://github.com/pytorch/pytorch/pull/124753)

For the record: once this PR is in we should also remove the jobs from the "unstable" workflow

File: .github/workflows/_check-user.yml
Comment (DanilBaibak) [reply .github/workflows/_check-user.yml:21]: I like the `runner determinator` version.

File: .github/workflows/_check-user.yml
Comment (DanilBaibak) [reply .github/workflows/_check-user.yml:18]: I general, I agree with you - we can use this approach dynamically selecting a runner. We don't need this feature at the moment. Let's keep the first version light and then iteratively improve it.

File: .github/workflows/pull.yml
Comment (DanilBaibak) [reply .github/workflows/pull.yml:43]: Can we test cron jobs at PR level?

**Comment by DanilBaibak** [comment]
As for https://github.com/pytorch/pytorch/pull/124721#discussion_r1585037322
I ran test workflows as dependencies. Or do you mean something else?

File: .github/workflows/_check-user.yml
Comment (DanilBaibak) [reply .github/workflows/_check-user.yml]: Yes, it won't work without it.

File: .github/workflows/_check-user.yml
Comment (DanilBaibak) [reply .github/workflows/_check-user.yml]: But would case insensitivity be worse for usernames?

File: .github/workflows/_check-user.yml
Comment (DanilBaibak) [reply .github/workflows/_check-user.yml]: My main idea is to make the solution simple. With a single list of people it is quite straightforward:
- If you are in the list, you use ARC, if not, then you don’t.
- You can also turn it on for everyone or turn it off.

There is no goal to maintain this system forever. The goal is to stabilize the ARC runners and only use them.

File: .github/workflows/pull.yml
Comment (DanilBaibak) [reply .github/workflows/pull.yml]: What do you mean by "test workflows"? The "_linux-test" workflows will be triggered after the corresponding build has completed:
Screenshot 2024-05-06 at 10 16 36

File: .github/workflows/_check-user.yml
Comment (DanilBaibak) [reply .github/workflows/_check-user.yml]: 1. I updated the code so the fallback policy always returns "`label`".
2. The [issue number](https://github.com/pytorch/pytorch/pull/124721/files#diff-5900a66fe6f5f773249ace156f682e6f2ef3618c6f8b536438e2519c2240eb53R17) is an input parameter. Thus, it should be easy to define and configure.

File: .github/scripts/get_workflow_type.py
Comment (jeanschmidt) [inline .github/scripts/get_workflow_type.py:43]: Lets not hardcode issue number. A better approach would be pick a issue, from a author list (ppl from our team) and with specific title and/or tag.

Hardcoding does not seems the way to go here.

File: .github/scripts/get_workflow_type.py
Comment (jeanschmidt) [inline .github/scripts/get_workflow_type.py:47]: very minor one:

```
return branch.split("/")[0] in {"main", "nightly", "release", "landchecks"}
```

File: .github/workflows/pull.yml
Comment (DanilBaibak) [reply .github/workflows/pull.yml:43]: I tested this separately. Running by a cron job, `github.head_ref` will return `null` and `github.ref_name` - branch name:
Screenshot 2024-05-06 at 13 56 00

**Review by jeanschmidt — APPROVED** [review]

File: .github/scripts/get_workflow_type.py
Comment (DanilBaibak) [reply .github/scripts/get_workflow_type.py:43]: Agreed to do as follow up improvement.

File: .github/workflows/_check-user.yml
Comment (ZainRizvi) [reply .github/workflows/_check-user.yml]: Github usernames are not actually case sensitive:
https://stackoverflow.com/questions/58591331/how-to-change-the-case-of-my-githubs-username-without-changing-any-letter#:~:text=Usernames%20are%20not%20case%20sensitive,that%20Username%20is%20not%20available.

File: .github/workflows/pull.yml
Comment (ZainRizvi) [inline .github/workflows/pull.yml:136]: Were you planning to uncomment and convert all of these workflows?

File: .github/scripts/get_workflow_type.py
Comment (ZainRizvi) [inline .github/scripts/get_workflow_type.py:90]: I'd like to push back on this and suggest we at least return a json dictionary for the response (e.g. `{"workflow_mode" : "both"}`)

Why? To preserve backwards compatibility. 

Because of the weird bake-time issues we've run into with GH Actions, we want to make sure that in the future (i.e. 2 weeks later) if we want to extend this method to return more granular information then we don't want to break anyone.

Having a dict will prevent us from having to worry about bake time first since we'll be able to keep returning a `"workflow_mode"` key for users on old forks, while adding new keys for the newer builds

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
Will approve once pull.yaml file is fixed (I'm guessing it wasn't done because this PR is still in draft mode)

**Pull request closed without merging**
